be1-go/cli/witness/witness.go

```diff
@@ -41,38 +41,80 @@ func Serve(context *cli.Context) error {
 		return xerrors.Errorf("failed to unmarshal public key: %v", err)
 	}
 
+	// create organizer hub
 	h := hub.NewWitnessHub(point)
 
-	ws, err := connectToOrganizer(organizerAddr, organizerPort)
+	// make context release resources associated with it when all operations are done
+	ctx, cancel := context.WithCancel(cliCtx.Context)
+	defer cancel()
+
+	// create wait group which waits for goroutines to finish
+	wg := &sync.WaitGroup{}
+
+	// increment wait group and connect to organizer's witness server
+	err = connectToSocket(ctx, hub.OrganizerSocketType, organizerAddress, h, organizerPort, wg)
 	if err != nil {
 		return xerrors.Errorf("failed to connect to organizer: %v", err)
 	}
 
-	organizerSocket := hub.NewOrganizerSocket(h, ws)
+	// increment wait group and connect to other witnesses
+	for _, otherWit := range otherWitness {
+		err = connectToSocket(ctx, hub.WitnessSocketType, otherWit, h, organizerPort, wg)
+		if err != nil {
+			return xerrors.Errorf("failed to connect to witness: %v", err)
+		}
+	}
 
-	go organizerSocket.WritePump()
-	go organizerSocket.ReadPump()
+	// increment wait group and create and serve servers for witnesses and clients
+	clientSrv := hub.CreateAndServeWS(ctx, hub.WitnessHubType, hub.ClientSocketType, h, clientPort, wg)
+	witnessSrv := hub.CreateAndServeWS(ctx, hub.WitnessHubType, hub.WitnessSocketType, h, witnessPort, wg)
 
-	done := make(chan struct{})
-	go h.Start(done)
+	// shut down client server and witness server when ctrl+c received
+	go hub.ShutDownServers(ctx, cancel, clientSrv, witnessSrv, wg)
```

Comment:
This should probably be after `go h.Start(ctx, wg)` and can be a blocking call i.e. without the `go` keyword.